src/Data/Monoid/DecidablyEmpty.hs

```diff
@@ -0,0 +1,115 @@
+instance DecidablyEmpty ()
+instance DecidablyEmpty Any
+instance DecidablyEmpty All
+-- instance DecidablyEmpty Lifetime
+-- instance DecidablyEmpty Event
+instance DecidablyEmpty [a] where
+  isEmpty = null
+instance
+#if MIN_VERSION_base(4,11,0)
+  Semigroup a
+#else
+  Monoid a
+#endif
+  => DecidablyEmpty (Maybe a) where
+  isEmpty = isNothing
+deriving instance (Num a, DecidablyEmpty a) => DecidablyEmpty (Product a)
+deriving instance (DecidablyEmpty a, Num a) => DecidablyEmpty (Sum a)
+deriving instance DecidablyEmpty a => DecidablyEmpty (Dual a)
+instance DecidablyEmpty (First a) where
+  isEmpty (First a) = isNothing a
+instance DecidablyEmpty (Last a) where
+  isEmpty (Last a) = isNothing a
+deriving instance DecidablyEmpty a => DecidablyEmpty (Identity a)
+instance Semigroup a => DecidablyEmpty (Option a) where
+  isEmpty (Option a) = isNothing a
+deriving instance DecidablyEmpty m => DecidablyEmpty (WrappedMonoid m)
+instance (Ord a, Bounded a) => DecidablyEmpty (Max a)
+instance (Ord a, Bounded a) => DecidablyEmpty (Min a)
+instance DecidablyEmpty (Proxy s)
+deriving instance DecidablyEmpty a => DecidablyEmpty (Const a b)
+#if MIN_VERSION_base(4,11,0)
+deriving instance DecidablyEmpty a => DecidablyEmpty (Down a)
+#endif
+#if MIN_VERSION_base(4,12,0)
+deriving instance DecidablyEmpty p => DecidablyEmpty (Par1 p)
+instance DecidablyEmpty (U1 p)
+deriving instance DecidablyEmpty (f p) => DecidablyEmpty (Rec1 f p)
+deriving instance DecidablyEmpty (f p) => DecidablyEmpty (M1 i c f p)
+deriving instance DecidablyEmpty c => DecidablyEmpty (K1 i c p)
+instance (DecidablyEmpty (f p), DecidablyEmpty (g p)) => DecidablyEmpty ((f :*: g) p) where
+  isEmpty (x :*: y) = isEmpty x && isEmpty y
+deriving instance DecidablyEmpty (f (g p)) => DecidablyEmpty ((f :.: g) p)
+#endif
+
+instance (DecidablyEmpty a, DecidablyEmpty b) => DecidablyEmpty (a, b) where
+  isEmpty (a, b) = isEmpty a && isEmpty b
+instance (DecidablyEmpty a, DecidablyEmpty b, DecidablyEmpty c) => DecidablyEmpty (a, b, c) where
+  isEmpty (a, b, c) = isEmpty a && isEmpty b && isEmpty c
+instance (DecidablyEmpty a, DecidablyEmpty b, DecidablyEmpty c, DecidablyEmpty d) => DecidablyEmpty (a, b, c, d) where
+  isEmpty (a, b, c, d) = isEmpty a && isEmpty b && isEmpty c && isEmpty d
+instance (DecidablyEmpty a, DecidablyEmpty b, DecidablyEmpty c, DecidablyEmpty d, DecidablyEmpty e) => DecidablyEmpty (a, b, c, d, e) where
+  isEmpty (a, b, c, d, e) = isEmpty a && isEmpty b && isEmpty c && isEmpty d && isEmpty e
+
+-- containers
+
+instance DecidablyEmpty IntSet.IntSet where
+  isEmpty = IntSet.null
+instance DecidablyEmpty (IntMap.IntMap v) where
+  isEmpty = IntMap.null
+instance Ord k => DecidablyEmpty (Map.Map k v) where
+  isEmpty = Map.null
+instance DecidablyEmpty (Seq.Seq v) where
+  isEmpty = Seq.null
+instance Ord k => DecidablyEmpty (Set.Set k) where
+  isEmpty = Set.null
+
+-- dependent-map
+
+instance GCompare k => DecidablyEmpty (DMap.DMap k v) where
+  isEmpty = DMap.null
```

Comment:
@danbornside points out these could use `Foldable`'s `null`; and that might be the better default method too.

(At the very least, we can provide newtypes for both for `deriving via`.)